### PR TITLE
Add regex support to wait for pods to be in Ready state

### PR DESCRIPTION
**Pull request by n3ph**
## Current Sitution

The tool is currently only capable of waiting for a specific amount of time between draining nodes ([`BETWEEN_NODES_WAIT`](https://wiki.commercetools.com/pages/viewpage.action?pageId=231605373)).

As we have some persistence layer around in K8s (MongoDB, ElasticSearch), we mostly ran into reoccurring issues when applying updates to the node groups. If a node will be drained all workload will be moved to a node in `Ready` state. Clusters like MongoDB or ElasticSearch have a very strict cluster health status and since one pod being moved to another node leads to a degraded cluster, we need some way to ensure all related pods are in `Ready` state (green cluster health status) before draining the next k8s worker node.

This PR aims to add support for that.

relates to https://github.com/commercetools/sre-cumulus/issues/349
relates to https://github.com/commercetools/sre-cumulus/issues/360

### CODE REVIEW

File: eksrollup/cli.py
Comment (swapzero) [inline eksrollup/cli.py]: This will still sleep `between_nodes_wait` seconds until the selected pods are ready and will do this multiple times. Maybe a while true with a `break` would be more suitable here.

Comment (n3ph) [reply eksrollup/cli.py]: this is intended.. But I can set that to a fixed value?
Reason is I do not want to DoS the K8s API

Comment (swapzero) [reply eksrollup/cli.py]: Yes, but this implementation cancels the idea of `between_nodes_wait` because there is the possibility that it will wait more than once if the pods are not in ready state

Comment (n3ph) [reply eksrollup/cli.py]: So what are you suggesting? a static timeout like 5 seconds?

Comment (swapzero) [reply eksrollup/cli.py]: Maybe these two options should mutually exclusive and the pods_ready check should have its own hardcoded timeout, something like 20 seconds, much less than between_nodes_wait

Comment (n3ph) [reply eksrollup/cli.py:241]: I`ll implement `BETWEEN_NODES_WAIT_POD_REGEX_WAIT` with a default vaule of 30 seconds.

Comment (n3ph) [reply eksrollup/cli.py]: Done by https://github.com/commercetools/eks-rolling-update/pull/1/commits/843eacfec5a6dbfa396f9f3923ee6be94ab6042b

File: eksrollup/cli.py
Comment (swapzero) [inline eksrollup/cli.py]: You also need to bail out at some point because this will loop forever if the pods are _never_ ready.

Comment (n3ph) [reply eksrollup/cli.py]: That is correct, IDK when... Suggestions?

Comment (swapzero) [reply eksrollup/cli.py]: have a `retries` variable with a value of 20 let's say. That would induce a delay of 600 seconds **top**, unless pods are ready earlier. If the retries has reached 0, get out of loop and  exit the script...smth like this

Comment (n3ph) [reply eksrollup/cli.py]: What about pods like ElasticSearch needing more time?

Comment (swapzero) [reply eksrollup/cli.py]: ¯\\_(ツ)_/¯ - retries = 60 (1800 seconds)

Comment (ktrzonkowski) [reply eksrollup/cli.py]: i don't know if this is overthinking... possible to maybe have it prompt user input of "wait for response or bail out" after running out of retries?  that may be silly, but it would let us be like "no no this is elasticsearch it's just gonna take a while... keep trying buddy..." 😂

Comment (n3ph) [reply eksrollup/cli.py]: I am willing to wait indefinitely... Since we would be informed about the pod we wait on, we would be able to consider abort anyway at any time...

**Comment by n3ph** [comment]
Build docker image with: `make docker-dist version=pod_regex_wait`

**Review by zonorti — APPROVED** [review]

**Review by swapzero — APPROVED** [review]

**Comment by swapzero** [comment]
I just realised that this _may_ fail if we don't specify a regex like `.*` because in the case of AWS setup, the only separated ASGs are for mongodb. We have two types of nodes: `nodes` and `mongodb`, so other deployments like backend ES run on general nodes and we can't rely on `search\d\d-` pattern in this case.

**Comment by n3ph** [comment]
> I just realised that this _may_ fail if we don't specify a regex like `.*` because in the case of AWS setup, the only separated ASGs are for mongodb. We have two types of nodes: `nodes` and `mongodb`, so other deployments like backend ES run on general nodes and we can't rely on `search\d\d-` pattern in this case.

No we don't. The code only looks for pods matching the regex, so if we would like to match mongodb and elastic elasticsearch we would use a regex like `^mongodb|^elasticsearch`. All other pods not matching the regex will be skipped by this check. The ASG does not matter, AFAIK.

**Comment by swapzero** [comment]
Yes, it does matter. here's an example:
if you use `(mongodb|elasticsearch)` regex and run the script on the `nodes` ASGs (ES is there as well), it will return immediately if the drained node **does not run** any `search....` and PDBs could be broken or a deployment could have no replicas.

**Comment by n3ph** [comment]
So you need to define a regex which matches any pods having PDBs defined, or what am I missing here?

File: eksrollup/cli.py
Comment (swapzero) [inline eksrollup/cli.py]: This code **is not executed** if ASG_USE_TERMINATION_POLICY is used...and guess what, I was using it for each run to avoid long waiting times until all instances are destroyed one by one...

LE: and it also fails atm, I will add a quick fix

File: eksrollup/lib/k8s.py
Comment (n3ph) [inline eksrollup/lib/k8s.py:274]: Please do not use short forms (`p`), use meaningful names instead

```suggestion
        pod_dict = pod.to_dict()
```

**Comment by svenmueller** [comment]
Waiting for an ES data K8s pod to be ready will not change the current described behavior (ELK cluster health status turning red under some circumstances). Even if the pod is ready, the ES cluster state might still be in health state "yellow" which means not all shard replicas are assigned yet (takes time). If then the next node is migrated, chances are that a shard primary and replica (note: we only have a single replica) are both not available (shard replica not fully assigned yet and primary not available due to node migration).

Waiting for ES cluster state to be "green" would help in that case.

**Comment by n3ph** [comment]
> Waiting for ES cluster state to be "green" would help in that case.

I thought the readiness probe is waiting for the cluster health status to be green... 😰

**Comment by svenmueller** [comment]
> > Waiting for ES cluster state to be "green" would help in that case.
> 
> I thought the readiness probe is waiting for the cluster health status to be green... 😰

Correct, there are some ES clusters (e.g. `pim-indexer`) who have customized the readiness probe ([example](https://github.com/commercetools/pim-indexer/blob/master/deployment/elasticsearch/helm-es7/elasticsearch/templates/statefulset.yaml#L223)). But that's not the case for all ES (e.g. our ELK currently checks if the API responds: [example](https://github.com/helm/charts/blob/master/stable/elasticsearch/values.yaml#L183-L186)). The readiness probe could be adjusted here (to use something like the pim-indexer ES).  We would also need to check the backend ES (search) regarding readiness probe logic.

**Comment by svenmueller** [comment]
@n3ph What's the status of this PR draft?

**Comment by n3ph** [comment]
I would say it's up to @commercetools/sre-cumulus to decide what to do with it.

**Pull request merged**
